pkg/wal/wal.go

```
@@ -42,19 +75,759 @@ type Segment interface {
+	if !log.entryCloser.AddRunning() {
+		return nil, errors.New("WAL is closed")
+	}
+	defer log.entryCloser.Done()
+
+	log.segmentMapMutex.RLock()
+	defer log.segmentMapMutex.RUnlock()
+
+	segments := make([]Segment, 0)
+	for _, segment := range log.segmentMap {
+		segments = append(segments, segment)
+	}
+	return segments, nil
+}
+
+// Rotate closes the open segment and opens a new one, returning the closed segment details.
+func (log *log) Rotate() (Segment, error) {
+	if !log.entryCloser.AddRunning() {
+		return nil, errors.New("WAL is closed")
+	}
+	defer log.entryCloser.Done()
+
+	oldWorkSegment, err := log.swapWorkSegment()
+	if err != nil {
+		return nil, err
+	}
+
+	log.segmentMapMutex.Lock()
+	defer log.segmentMapMutex.Unlock()
+
+	// Update segment information.
+	workSegment := log.workSegment
+	log.segmentMap[workSegment.segmentID] = workSegment
+	return oldWorkSegment, nil
+}
+
+// Delete the specified segment.
+func (log *log) Delete(segmentID SegmentID) error {
+	if !log.entryCloser.AddRunning() {
+		return errors.New("WAL is closed")
+	}
+	defer log.entryCloser.Done()
+
+	log.segmentMapMutex.Lock()
+	defer log.segmentMapMutex.Unlock()
+
+	// Segment which will be deleted must be closed.
+	if segmentID == log.workSegment.segmentID {
+		return errors.New("Can not delete the segment which is working")
+	}
+
+	err := os.Remove(log.segmentMap[segmentID].path)
+	if err != nil {
+		return errors.Wrap(err, "Delete WAL segment error")
+	}
+	delete(log.segmentMap, segmentID)
+	return nil
+}
+
+// Close all of segments and stop WAL work.
+func (log *log) Close() error {
+	log.closerOnce.Do(func() {
+		log.logger.Info().Msg("Closing WAL...")
+
+		log.entryCloser.Done()
+		log.entryCloser.CloseThenWait()
+
+		close(log.writeChannel)
+		log.writeWaitGroup.Wait()
+
+		close(log.flushChannel)
+		log.flushWaitGroup.Wait()
+
+		if err := log.flushBuffer(log.buffer); err != nil {
+			log.logger.Err(err).Msg("Flushing buffer failed")
```

Please return the error more than log it.